python/pyiceberg/table/__init__.py

```diff
@@ -934,206 +976,630 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
-            except ValueError:
-                pass
+    def _set_column_requirement(self, path: Union[str, Tuple[str, ...]], required: bool) -> None:
+        path = (path,) if isinstance(path, str) else path
+        name = ".".join(path)
 
-            if exist_field:
-                raise ValueError(f"Cannot add column, name already exists: {parent}.{name}")
+        field = self._schema.find_field(name, self._case_sensitive)
 
-            full_name = parent_field.name + "." + name
+        if (field.required and required) or (field.optional and not required):
+            # if the change is a noop, allow it even if allowIncompatibleChanges is false
+            return
 
+        if not self._allow_incompatible_changes and required:
+            raise ValueError(f"Cannot change column nullability: {name}: optional -> required")
+
+        if field.field_id in self._deletes:
+            raise ValueError(f"Cannot update a column that will be deleted: {name}")
+
+        if updated := self._updates.get(field.field_id):
+            self._updates[field.field_id] = NestedField(
+                field_id=updated.field_id,
+                name=updated.name,
+                field_type=updated.field_type,
+                doc=updated.doc,
+                required=required,
+            )
         else:
-            try:
-                exist_field = self._schema.find_field(name, self._case_sensitive)
-            except ValueError:
-                pass
+            self._updates[field.field_id] = NestedField(
+                field_id=field.field_id,
+                name=field.name,
+                field_type=field.field_type,
+                doc=field.doc,
+                required=required,
+            )
+
+    def update_column(
+        self,
+        path: Union[str, Tuple[str, ...]],
+        field_type: Optional[IcebergType] = None,
+        required: Optional[bool] = None,
+        doc: Optional[str] = None,
+    ) -> UpdateSchema:
+        """Update the type of column.
 
-            if exist_field:
-                raise ValueError(f"Cannot add column, name already exists: {name}")
+        Args:
+            path: The path to the field.
+            field_type: The new type
+            required: If the field should be required
+            doc: Documentation describing the column
 
-        # assign new IDs in order
-        new_id = self.assign_new_column_id()
+        Returns:
+            The UpdateSchema with the type update staged.
+        """
+        path = (path,) if isinstance(path, str) else path
+        full_name = ".".join(path)
+
+        field = self._schema.find_field(full_name, self._case_sensitive)
+
+        if field.field_id in self._deletes:
+            raise ValueError(f"Cannot update a column that will be deleted: {full_name}")
+
+        if field_type is not None:
+            if not self._allow_incompatible_changes and field.field_type != field_type:
```

Comment:
The Java equivalent ensures that only primitive fields are updated because the only type that can be passed in is a `PrimitiveType`. I think we need to have an additional check here that this is not replacing a nested type.

It may be that `promote` already does this for us, but I want to highlight that we cannot allow calling this to update a `struct`, `map`, or `list` type.

Reply:
Schema evolution catches this:
```
Cannot change column type: foo: struct<2: bar: optional string> -> string
```
I've added a check to make the error more specific error:
```
Cannot change column type: struct<2: bar: optional string> is not a primitive
```